### PR TITLE
Improve ODE example to output at multiple times, and make better use …

### PR DESCRIPTION
**Pull request by duvenaud**
…of types.

Separating the solver into 4 functions is overkill for Euler's method, but lays the groundwork for an adaptive-step RK solver.

Co-authored by: @dougalm

### CODE REVIEW

File: examples/ode-integrator.dx [hunk 1]
```diff
@@ -2,40 +2,56 @@
 Using only generic vector operations.
 
 Time : Type = Real
+def Eulerstate (a:Type) : Type = (a & Time & a & Time)
```
Comment (apaszke) [inline examples/ode-integrator.dx:5]: docs pls? 😊

Comment (duvenaud) [reply examples/ode-integrator.dx:7]: Added an explanation - is this what you had in mind?

Comment (apaszke) [reply examples/ode-integrator.dx:7]: Yeah, perfect. Thanks!

**Comment by duvenaud** [comment]
> This LGTM, but I can't merge due to conflicts!

Whoops, fixed now.

**Pull request merged**